src/uct/ib/base/ib_md.c

```diff
@@ -303,149 +292,6 @@ uct_ib_md_print_mem_reg_err_msg(const char *title, void *address, size_t length,
-    size_t length;
-
-    while (ctx->length > 0) {
-        length = ucs_min(ctx->length, max_chunk);
-        if (ctx->params != NULL) {
-            status = uct_ib_reg_mr_params(ctx->md, ctx->address, length,
-                                          ctx->params, ctx->access_flags,
-                                          &ctx->mrs[mr_idx]);
-            if (status != UCS_OK) {
-                goto err_dereg;
-            }
-        } else {
-            status = uct_ib_dereg_mr(ctx->mrs[mr_idx]);
-            if (status != UCS_OK) {
-                goto err;
-            }
-        }
-        ctx->address = UCS_PTR_BYTE_OFFSET(ctx->address, length);
-        ctx->length -= length;
-        mr_idx++;
-    }
-
-    ucs_trace("%s %p..%p took %f usec\n", ctx->reg ? "reg_mr" : "dereg_mr",
-              ctx->mrs[0]->addr, ctx->address,
-              ucs_time_to_usec(ucs_get_time() - t0));
-    return UCS_STATUS_PTR(UCS_OK);
-
-err_dereg:
-    for (; mr_idx >= 0; --mr_idx) {
-        uct_ib_dereg_mr(ctx->mrs[mr_idx]);
-    }
-err:
-    return UCS_STATUS_PTR(status);
-}
-
-ucs_status_t
-uct_ib_md_handle_mr_list_mt(uct_ib_md_t *md, void *address, size_t length,
-                            const uct_md_mem_reg_params_t *params,
-                            uint64_t access_flags, struct ibv_mr **mrs)
-{
-    size_t chunk_size = md->config.mt_reg_chunk;
-    int mr_num        = ucs_div_round_up(length, chunk_size);
-    int thread_num_mrs, thread_num, thread_idx, mr_idx, cpu_id;
-    ucs_sys_cpuset_t parent_set, thread_set;
-    uct_ib_md_mem_reg_thread_t *ctxs, *ctx;
-    char UCS_V_UNUSED affinity_str[64];
-    pthread_attr_t attr;
-    ucs_status_t status;
-    void *thread_status;
-    int ret;
-
-    status = ucs_sys_pthread_getaffinity(&parent_set);
-    if (status != UCS_OK) {
-        return status;
-    }
-
-    thread_num = ucs_min(CPU_COUNT(&parent_set), mr_num);
-    if (thread_num == 1) {
-        return UCS_ERR_UNSUPPORTED;
-    }
-
-    ucs_trace("multithreaded %s %p..%p threads %d affinity %s\n",
-              (params != NULL) ? "reg" : "dereg", address,
-              UCS_PTR_BYTE_OFFSET(address, length), thread_num,
-              ucs_make_affinity_str(&parent_set, affinity_str,
-                                    sizeof(affinity_str)));
-
-    ctxs = ucs_calloc(thread_num, sizeof(*ctxs), "ib mr ctxs");
-    if (ctxs == NULL) {
-        return UCS_ERR_NO_MEMORY;
-    }
-
-    pthread_attr_init(&attr);
-
-    status = UCS_OK;
-    mr_idx = 0;
-    cpu_id = 0;
-    for (thread_idx = 0; thread_idx < thread_num; thread_idx++) {
-        /* calculate number of mrs for each thread so each one will
-         * get proportional amount */
-        thread_num_mrs    = ucs_div_round_up(mr_num - mr_idx,
-                                             thread_num - thread_idx);
-        ctx               = &ctxs[thread_idx];
-        ctx->md           = md;
-        ctx->address      = UCS_PTR_BYTE_OFFSET(address, mr_idx * chunk_size);
```

Comment:
can't we calculate the address of each mr[I] like this instead of saving all of them in the ksm_data array?

Reply:
Good idea, thanks! Done.

BTW, I moved several functions (`uct_ib_md_handle_mr_list_mt` and `uct_ib_md_mem_handle_thread_func`) from `ib_md.c` to `ib_mlx5dv_md.c` since they were used only if DEVX is enabled. WDYT, should I move them back?